Regression: CyclomaticComplexity.switchBlockAsSingleDecisionPoint does not work properly anymore (10.21.1)

I have read check documentation: https://checkstyle.org/checks/metrics/cyclomaticcomplexity.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java
Successfully compiled with JDK 21

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="CyclomaticComplexity">
            <property name="max" value="3"/>
            <property name="switchBlockAsSingleDecisionPoint" value="true"/>
        </module>
    </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    // Passed: Cyclomatic Complexity is 2
    void test(Object obj) { // 1, method declaration
        switch (obj) { // 2, switch
            case Integer i -> System.out.println("Integer: " + i);
            case String s -> System.out.println("String " + s);
            default -> System.out.println("none");
        }
    }

    // Violation: Cyclomatic Complexity is 4, but must be 2 since
    // switchBlockAsSingleDecisionPoint = true and "when" literals must not be counted
    void test2(Object obj1, Object obj2) { // 1, method declaration
        switch (obj1) { // 2, switch
            case Integer i1 when obj2 instanceof Integer i2 -> System.out.println("Integers: %d, %d".formatted(i1, i2));
            case String s1 when obj2 instanceof String s2 -> System.out.println("Strings: %s, %s".formatted(s1, s2));
            default -> System.out.println("none");
        }
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.21.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:13:5: Cyclomatic Complexity is 4 (max allowed is 3). [CyclomaticComplexity]
Audit done.
Checkstyle ends with 1 errors.
```
---

switchBlockAsSingleDecisionPoint is set to true in config.xml, so switch block should be treated as a single decision point.
However, in case of "test2" method "when" literals are counted, when must not.
This is regression after fix of #15341

---

I am ok to approve.

I brought up this at (https://github.com/checkstyle/checkstyle/pull/15341#discussion_r1686490907) but it looks like we chose to keep the behavior consistent with other cases

Example: 

Replace `when` with `if`.

```
D:\CS\test
cat src/Test.java
public class Test {
    void test2(Object obj1, Object obj2) { // 1, method declaration
        switch (obj1) { // 2, switch
            case Integer i1 -> {
                if (obj2 instanceof Integer i2)   // 3, if  ( not ignored)
                    System.out.println("Integers: %d, %d".formatted(i1, i2));
            }
            case String s1 -> {
                if (obj2 instanceof String s2)  // 4, if  ( not ignored)
                    System.out.println("Strings: %s, %s".formatted(s1, s2));
            }
            default -> System.out.println("none");
        }
    }
}
D:\CS\test
java  -jar checkstyle-10.21.1-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:2:5: Cyclomatic Complexity is 4 (max allowed is 3). [CyclomaticComplexity]
Audit done.
Checkstyle ends with 1 errors.

```

The current logic in the implementation increments the complexity for any token inside the switch block except case statements, even when `switchBlockAsSingleDecisionPoint` is true (https://github.com/checkstyle/checkstyle/pull/15341#discussion_r1691323211)


But IMO, the behavior described in the issue is inconsistent with the documented behavior of the CyclomaticComplexity check when the `switchBlockAsSingleDecisionPoint` property is true. According to the documentation and the intent of this property, the entire switch block should be treated as a single decision point, 

> switchBlockAsSingleDecisionPoint: Control whether to treat the whole switch block as a single decision point.

https://checkstyle.sourceforge.io/checks/metrics/cyclomaticcomplexity.html

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/CyclomaticComplexityCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/CyclomaticComplexityCheck.java
@@ -27,6 +27,7 @@ import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
 import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
 
 /**
  * <div>
@@ -270,7 +271,7 @@ public class CyclomaticComplexityCheck
      */
     private void visitTokenHook(DetailAST ast) {
         if (switchBlockAsSingleDecisionPoint) {
-            if (ast.getType() != TokenTypes.LITERAL_CASE) {
+            if (!ScopeUtil.isInBlockOf(ast, TokenTypes.LITERAL_SWITCH)) {
                 incrementCurrentValue(BigInteger.ONE);
             }
         }

--- a/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/utils/ScopeUtil.java
@@ -199,7 +199,7 @@ public final class ScopeUtil {
      * @param tokenType type of token.
      * @return a {@code boolean} value
      */
-    private static boolean isInBlockOf(DetailAST node, int tokenType) {
+    public static boolean isInBlockOf(DetailAST node, int tokenType) {
         boolean returnValue = false;
 
         // Loop up looking for a containing interface block

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/CyclomaticComplexityCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/CyclomaticComplexityCheckTest.java
@@ -182,13 +182,25 @@ public class CyclomaticComplexityCheckTest
     public void testWhenExpressionSwitchAsSinglePoint() throws Exception {
         final String[] expected = {
             "14:5: " + getCheckMessage(MSG_KEY, 5, 0),
-            "20:5: " + getCheckMessage(MSG_KEY, 4, 0),
-            "29:5: " + getCheckMessage(MSG_KEY, 5, 0),
-            "39:5: " + getCheckMessage(MSG_KEY, 5, 0),
+            "20:5: " + getCheckMessage(MSG_KEY, 2, 0),
+            "29:5: " + getCheckMessage(MSG_KEY, 2, 0),
+            "39:5: " + getCheckMessage(MSG_KEY, 2, 0),
         };
         verifyWithInlineConfigParser(
                 getNonCompilablePath(
                         "InputCyclomaticComplexityWhenSwitchAsSinglePoint.java"), expected);
+    }
+
+    @Test
+    public void testSwitchBlockAsSingleDecisionPointWithNestedSwitch() throws Exception {
+        final String[] expected = {
+            "17:5: " + getCheckMessage(MSG_KEY, 2, 0),
+            "26:5: " + getCheckMessage(MSG_KEY, 2, 0),
+            "41:5: " + getCheckMessage(MSG_KEY, 2, 0),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputCyclomaticComplexitySwitchBlocks6.java"), expected);
     }
 
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexitySwitchBlocks6.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexitySwitchBlocks6.java
@@ -1,0 +1,75 @@
+/*
+CyclomaticComplexity
+max = 0
+switchBlockAsSingleDecisionPoint = true
+tokens = (default)LITERAL_WHILE, LITERAL_DO, LITERAL_FOR, LITERAL_IF, LITERAL_SWITCH, \
+         LITERAL_CASE, LITERAL_CATCH, QUESTION, LAND, LOR
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.metrics.cyclomaticcomplexity;
+
+public class InputCyclomaticComplexitySwitchBlocks6 {
+
+    // violation below, 'Cyclomatic Complexity is 2 (max allowed is 0)'
+    void test1(Object obj1, Object obj2) { // 1, method declaration
+        switch (obj1) { // 2, switch
+            case Integer i1 when obj2 instanceof Integer i2 -> System.out.println("1");
+            case String s1 when obj2 instanceof String s2 -> System.out.println("2");
+            default -> System.out.println("none");
+        }
+    }
+
+    // violation below, 'Cyclomatic Complexity is 2 (max allowed is 0)'
+    void test2(Object obj1, Object obj2) { // 1, method declaration
+        switch (obj1) { // 2, switch
+            case Integer i1 -> {
+                if (obj2 instanceof Integer i2)   // if ignored
+                    System.out.println("Integers: %d, %d".formatted(i1, i2));
+            }
+            case String s1 -> {
+                if (obj2 instanceof String s2)  // if ignored
+                    System.out.println("Strings: %s, %s".formatted(s1, s2));
+            }
+            default -> System.out.println("none");
+        }
+    }
+
+    // violation below, 'Cyclomatic Complexity is 2 (max allowed is 0)'
+    void test3(Object obj1, Object obj2, Object obj3) {
+        int j1 = 0, k1 = 0, l1 = 0, m1 = 0; // Initialized variables
+
+        switch (obj1) { // 1
+            case Integer i -> {
+                if (i == 1) { // ignored
+                    System.out.println("Integers");
+                }
+                else if (i == 2) {
+                    switch (obj2) { // ignored
+                        case Integer s1 -> System.out.println("Integers");
+                        default -> System.out.println("none");
+                    }
+                }
+                else if (i == 3) {
+                    if (j1 == 1) { // ignored
+                        k1 = 1;
+                    } else {
+                        l1 = 1;
+                    }
+                }
+                else if (i == 4) {
+                    switch (obj3) { // ignored
+                        case Integer s2 -> j1++;
+                        default -> System.out.println("none");
+                    }
+                }
+                else {
+                    System.out.println("none");
+                }
+            }
+            default -> System.out.println("none");
+        }
+    }
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexityWhenSwitchAsSinglePoint.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/metrics/cyclomaticcomplexity/InputCyclomaticComplexityWhenSwitchAsSinglePoint.java
@@ -16,36 +16,36 @@ public class InputCyclomaticComplexityWhenSwitchAsSinglePoint {
         else if (o instanceof String s && s.length() == 9) {} // 4, if - 5, &&
     }
 
-    // violation below, 'Cyclomatic Complexity is 4 (max allowed is 0)'
+    // violation below, 'Cyclomatic Complexity is 2 (max allowed is 0)'
     void testSwitch(Object o) {   // 1, function declaration
         switch (o) {  // 2, switch
-            case Integer i when i == 9 -> {} // 3, when
-            case String s when s.length() == 9 -> {} // 4, when
+            case Integer i when i == 9 -> {}
+            case String s when s.length() == 9 -> {}
             default -> {}
         }
     }
 
-    // violation below, 'Cyclomatic Complexity is 5 (max allowed is 0)'
+    // violation below, 'Cyclomatic Complexity is 2 (max allowed is 0)'
     void testSwitch2(Object o) {  // 1, function declaration
         switch (o) {    // 2 switch
             case Integer i -> {}
-            case Point(int x, int y) when (x == 0 && y == 0) -> {} // 3, when - 4, &&
-            case Point(int x, int y) when x == 0 -> {} // 5, when
+            case Point(int x, int y) when (x == 0 && y == 0) -> {} // &&
+            case Point(int x, int y) when x == 0 -> {}
             default -> {}
         }
     }
 
-    // violation below, 'Cyclomatic Complexity is 5 (max allowed is 0)'
+    // violation below, 'Cyclomatic Complexity is 2 (max allowed is 0)'
     void testSwitch3(Object o) {  // 1, function declaration
         switch (o) {    // 2 switch
             case Integer i -> {}
             case Point(int x, int y) -> {
-                if (x == 0 && y == 0) { // 3, if - 4, &&
+                if (x == 0 && y == 0) { // &&
 
                 }
             }
             case Point2(int x) -> {
-                if (x == 0) { // 5 if
+                if (x == 0) {
 
                 }
             }

